src/libostree/ostree-repo.c

```diff
@@ -4379,6 +4379,171 @@ ostree_repo_read_commit (OstreeRepo   *self,
+                                     GCancellable     *cancellable,
+                                     GError           **error);
+
+/**
+ * _ostree_repo_open_file:
+ * @self: Repo
+ * @ref: ref, commit checksum, tree checksum with optional :path suffix
+ * @path: subpath under ref to return.  Many of the command-line utilities take
+ *     a --subpath or similar argument.  This is provided as a convenience to
+ *     handle those.  Set to %NULL if not needed.
+ * @flags: Specify what kind of file the caller is expecting - a tree, a blob
+ *     or either.
+ * @out_root: (out): An #OstreeRepoFile corresponding to the tree/blob.
+ * @out_commit_sha: (out): SHA of the commit that the file belongs to if any.
+ *    If the commit is unknown *out will be set to %NULL.
+ * @cancellable: Cancellable
+ * @error: Error
+ *
+ * This is a helper for the command line utilities.  It's used to convert a
+ * reference to a tree or a file from the user into an OstreeRepoFile.  It takes
+ * care to provide good error messages.
+ *
+ * @ref can take the form:
+ *
+ *    REMOTE:REF
+ *    :REF
+ *    REF
+ *    COMMIT_SHA
+ *
+ * Not currently supported (but could be implemented):
+ *
+ *    BLOB_SHA
+ *
+ * Note: if specifying `TREE_SHA` PATH must not be empty otherwise we won't have
+ * a valid metadata checksum for the directory.
+ */
+gboolean
+_ostree_repo_open_file (OstreeRepo               *self,
+                        const char               *ref,
+                        const char               *path,
+                        OstreeRepoOpenFileFlags   flags,
+                        OstreeRepoFile          **out,
+                        char                    **out_commit_sha,
+                        GCancellable             *cancellable,
+                        GError                  **error)
+{
+  g_return_val_if_fail (ref, FALSE);
+
+  g_autofree char * rev = g_strdup (ref);
+  g_autofree char * commit_sha = NULL;
+
+  g_autofree char * sha = NULL;
+  /* ostree_repo_resolve_rev passes things that look like SHAs through, so we're
+   * safe passing tree SHAs in here. */
+  if (!ostree_repo_resolve_rev (self, rev, FALSE, &sha, error))
+    return FALSE;
+
+  g_autoptr (OstreeRepoFile) ret_out = NULL;
+  if (!resolve_sha_to_tree (self, sha, &ret_out, &commit_sha, cancellable,
+                            error))
+    return FALSE;
+
+  if (path && g_strcmp0 (path, "") != 0 && g_strcmp0 (path, "/") != 0 &&
+      g_strcmp0 (path, "/.") != 0)
+    {
+      g_set_object (&ret_out, OSTREE_REPO_FILE (
+            g_file_resolve_relative_path ((GFile*) ret_out, path)));
+      if (!ret_out)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                       "File \"%s\" not found", path);
+          return FALSE;
+        }
+    }
+
+  if (g_strcmp0 (ostree_repo_file_tree_get_metadata_checksum (ret_out),
+                 "00000000000000000000000000000000") == 0)
```

Comment:
Hmm...we don't use the all-zero sha256 bit pattern anywhere do we?  The empty tree metadata checksum looks like it's `6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d` (from `/usr/share/empty`) which is itself different from the empty sha256:

```
$ sha256sum </dev/null
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855  -
$
```

Am I misunderstanding something?  How can this case be hit?

Reply:
I set it explicitly below in `resolve_sha_to_tree`.